### ISSUE TITLE
Resource leak when destroying non-connected handle with associated callbacks 

### ISSUE THREAD

**Issue by forGGe** [eclipse-paho/paho.mqtt.c#976]
When Paho fails to connect to an MQTT endpoint, it keeps a socket open even after `MQTTClient_destroy` is called. `MQTTClient_setCallbacks` prior to `MQTTClient_connect` seems to be important here.

**To Reproduce**
1. Take the code below:

```c
#include <MQTTClient.h>
#include <stdlib.h>
#include <unistd.h>

static int placeholder_cb(void *context, char *topic_name, int topic_len, MQTTClient_message *msg)
{
    (void)context;
    (void)topic_name;
    (void)topic_len;
    (void)msg;
    return 0;
}

int main(void)
{
    int rc;

    for (;;) {
        MQTTClient handle = NULL;

        fprintf(stderr, ">>>> trying to create MQTT handle...\n");

        rc = MQTTClient_create(&handle,
                "localhost:1234",
                "test_id",
                MQTTCLIENT_PERSISTENCE_NONE,
                NULL);

        if (rc != MQTTCLIENT_SUCCESS) {
            fprintf(stderr, "failed to create client\n");
            exit(1);
        }

        fprintf(stderr, ">>>> setting callbacks\n");

        rc = MQTTClient_setCallbacks(handle, &handle, NULL, placeholder_cb, NULL);
        if (rc != MQTTCLIENT_SUCCESS) {
            fprintf(stderr, "failed to set callback\n");
            exit(1);
        }

        MQTTClient_connectOptions opts = MQTTClient_connectOptions_initializer;

        opts.username = "user",
        opts.password = "pass",

        fprintf(stderr, ">>>> connecting client\n");

        rc = MQTTClient_connect(handle, &opts);
        if (rc != MQTTCLIENT_SUCCESS) {
            fprintf(stderr, ">>>> failed to connect, keep going...\n");
        }

        MQTTClient_destroy(&handle);
        fprintf(stderr, ">>>> client handle destroyed\n");
        sleep(3);
    }
}
```

2. Compile it with the following line:

```
gcc -Wall -Wextra ~/projects/tektelic/mqtt-snmp-bridge/paho_test.c -lpaho-mqtt3cs
```

3. Run it under `strace`:

```
strace -e trace=socket,close ./a.out
```

4. In the separate console run:

```
watch lsof -p $(pidof a.out)
```

**Expected behavior**

`strace` shows `socket`/`close` clearly paired, the file descriptor number is not incremented after each run.
`lsof` shows a file list staying the same.

**Actual behavior**

* After each step, `strace` shows a socket being allocated, but not deleted:

(Pay attention to lines `socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 4` and `socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 6`)

```
>>>> trying to create MQTT handle...
>>>> setting callbacks
>>>> connecting client
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 3
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 3
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 3
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
socket(AF_INET6, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 3
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 4
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 4
socket(AF_INET6, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 4
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 4
>>>> failed to connect, keep going...
>>>> client handle destroyed

...

>>>> trying to create MQTT handle...
>>>> setting callbacks
>>>> connecting client
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 5
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 5
socket(AF_INET6, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 5
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 5
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 6
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 6
socket(AF_INET6, SOCK_DGRAM|SOCK_CLOEXEC, IPPROTO_IP) = 6
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 6
>>>> failed to connect, keep going...
>>>> client handle destroyed
``` 

* `lsof` shows a growing list of open FDs:

```
                                                                                                                     16:19:52
COMMAND   PID     USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
a.out   36791     user  cwd    DIR   0,47       60   231781 /tmp/paho_test
a.out   36791     user  rtd    DIR  254,0     4096        2 /
a.out   36791     user  txt    REG   0,47    17000   245262 /tmp/paho_test/a.out
a.out   36791     user  mem    REG  254,0   100376 16030310 /usr/lib/libnss_myhostname.so.2
a.out   36791     user  mem    REG  254,0   291432 16030312 /usr/lib/libnss_mymachines.so.2
a.out   36791     user  mem    REG  254,0    51376 15994285 /usr/lib/libnss_files-2.32.so
a.out   36791     user  mem    REG  254,0    18608 15994232 /usr/lib/libdl-2.32.so
a.out   36791     user  mem    REG  254,0  2986744 15999779 /usr/lib/libcrypto.so.1.1
a.out   36791     user  mem    REG  254,0   585320 15999782 /usr/lib/libssl.so.1.1
a.out   36791     user  mem    REG  254,0   158744 15994297 /usr/lib/libpthread-2.32.so
a.out   36791     user  mem    REG  254,0  2159552 15994214 /usr/lib/libc-2.32.so
a.out   36791     user  mem    REG  254,0   166504 16017304 /usr/lib/libpaho-mqtt3cs.so.1.0
a.out   36791     user  mem    REG  254,0   207944 15994198 /usr/lib/ld-2.32.so
a.out   36791     user    0u   CHR  136,2      0t0        5 /dev/pts/2
a.out   36791     user    1u   CHR  136,2      0t0        5 /dev/pts/2
a.out   36791     user    2u   CHR  136,2      0t0        5 /dev/pts/2
a.out   36791     user    3u  sock    0,8      0t0   255027 protocol: TCP
a.out   36791     user    4u  sock    0,8      0t0   252784 protocol: TCP
a.out   36791     user    5u  sock    0,8      0t0   253907 protocol: TCP
a.out   36791     user    6u  sock    0,8      0t0   253911 protocol: TCP
a.out   36791     user    7u  sock    0,8      0t0   252803 protocol: TCP
a.out   36791     user    8u  sock    0,8      0t0   256048 protocol: TCP
a.out   36791     user    9u  sock    0,8      0t0   256076 protocol: TCP
a.out   36791     user   10u  sock    0,8      0t0   254968 protocol: TCP
a.out   36791     user   11u  sock    0,8      0t0   256080 protocol: TCP
a.out   36791     user   12u  sock    0,8      0t0   256084 protocol: TCP
a.out   36791     user   13u  sock    0,8      0t0   257061 protocol: TCP
a.out   36791     user   14u  sock    0,8      0t0   252864 protocol: TCP
a.out   36791     user   15u  sock    0,8      0t0   252885 protocol: TCP
a.out   36791     user   16u  sock    0,8      0t0   252889 protocol: TCP
a.out   36791     user   17u  sock    0,8      0t0   255199 protocol: TCP
a.out   36791     user   18u  sock    0,8      0t0   255203 protocol: TCP
a.out   36791     user   19u  sock    0,8      0t0   258069 protocol: TCP
a.out   36791     user   20u  sock    0,8      0t0   255225 protocol: TCP
a.out   36791     user   21u  sock    0,8      0t0   258086 protocol: TCP
a.out   36791     user   22u  sock    0,8      0t0   258090 protocol: TCP
a.out   36791     user   23u  sock    0,8      0t0   258118 protocol: TCP
a.out   36791     user   24u  sock    0,8      0t0   258121 protocol: TCP
a.out   36791     user   25u  sock    0,8      0t0   256192 protocol: TCP
a.out   36791     user   26u  sock    0,8      0t0   256196 protocol: TCP
a.out   36791     user   27u  sock    0,8      0t0   257202 protocol: TCP
a.out   36791     user   28u  sock    0,8      0t0   258160 protocol: TCP
a.out   36791     user   29u  sock    0,8      0t0   257220 protocol: TCP
a.out   36791     user   30u  sock    0,8      0t0   257224 protocol: TCP
a.out   36791     user   31u  sock    0,8      0t0   257241 protocol: TCP
a.out   36791     user   32u  sock    0,8      0t0   256249 protocol: TCP
a.out   36791     user   33u  sock    0,8      0t0   255352 protocol: TCP
a.out   36791     user   34u  sock    0,8      0t0   255356 protocol: TCP
a.out   36791     user   35u  sock    0,8      0t0   258387 protocol: TCP
a.out   36791     user   36u  sock    0,8      0t0   258391 protocol: TCP

```

After some time the application spends all available FDs, and other components (if present) start throwing "Too much open file" errors.

** Environment (please complete the following information):**
 - OS: ArchLinux `5.8.14-arch1-1`, embedded Linux `3.12.17`
 - Paho v1.3.1.r30.gfbf9828-1, Paho v1.3.0

**Additional context**

For more debugging stuff, run the `strace` with following params:

```
strace -k -e trace=socket,close ./a.out
```

...which yields (in my env) following backtrace for the forever-open socket: 

```
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 4
 > /usr/lib/libc-2.32.so(socket+0xb) [0x1019ab]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTReasonCode_toString+0x1f71) [0x17111]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTProperties_getProperty+0x208d) [0x14c2d]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTClient_yield+0x575) [0x8d95]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTClient_yield+0x13b5) [0x9bd5]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTClient_yield+0x1949) [0xa169]
 > /usr/lib/libpaho-mqtt3cs.so.1.0(MQTTClient_connect+0x37) [0xa267]
 > /tmp/paho_test/a.out(main+0x21c) [0x13cf]
 > /usr/lib/libc-2.32.so(__libc_start_main+0xf2) [0x28152]
 > /tmp/paho_test/a.out(_start+0x2e) [0x10ce]
```

**Comment by icraggs** [comment]
With the latest release (1.3.6) I get:

    >>>> trying to create MQTT handle...
    close(3)                                = 0
    >>>> setting callbacks
    >>>> connecting client
    close(3)                                = 0
    socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
    close(3)                                = 0
    close(3)                                = 0
    socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
    close(3)                                = 0    
    >>>> failed to connect, keep going...
    >>>> client handle destroyed
    >>>> trying to create MQTT handle...
    close(3)                                = 0  
    >>>> setting callbacks
    >>>> connecting client
    close(3)                                = 0
    socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
    close(3)                                = 0
    close(3)                                = 0 
    socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 3
    close(3)                                = 0
    >>>> failed to connect, keep going...

Something like this was fixed in 1.3.2 as far as I recall.

**Comment by forGGe** [comment]
> Something like this was fixed in 1.3.2 as far as I recall.

Thanks for the quick answer!

I tried with v1.3.2 and, indeed, I can't see FDs leaking.
Closing this issue.
